src/ffi/mod.rs

```diff
@@ -838,6 +843,98 @@ extern "C" {
     fn fortify_source();
 }
 
+#[allow(unused)]
+struct LogCensor {
+    ip_mask_seed: [u8; 32],
+    ip_regex: Regex,
+}
+
+#[allow(unused)]
+impl LogCensor {
+    fn new() -> Result<Self, RegexError> {
+        RegexBuilder::new(
+            r"
+                (?:
+                    (?:
+                        25[0-5]
+                        |  2[0-4][0-9]
+                        |  1[0-9]{2}
+                        |  [1-9]?[0-9]
+                    )\.
+                ){3}
+                (?:
+                    25[0-5]
+                    |  2[0-4][0-9]
+                    |  1[0-9]{2}
+                    |  [1-9]?[0-9]
+                )
+                |
+                (?:
+                    ([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}
+                    |   :(:[0-9a-fA-F]{1,4}){1,7}
+                    |   ([0-9a-fA-F]{1,4}:){1}(:[0-9a-fA-F]{1,4}){1,6}
+                    |   ([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}
+                    |   ([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}
+                    |   ([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}
+                    |   ([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}
+                    |   ([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}
+                    |   ([0-9a-fA-F]{1,4}:){1,7}:
+                )
+                ",
+        )
+        .ignore_whitespace(true)
+        .build()
+        .map(|re| LogCensor {
+            ip_mask_seed: rand::thread_rng().gen::<[u8; 32]>(),
+            ip_regex: re,
+        })
+    }
+
+    fn censor_logs<'a>(&self, log: &'a str) -> std::borrow::Cow<'a, str> {
+        // Gather all of the IPs
+        self.ip_regex.replace_all(log, |captures: &Captures| {
+            captures
+                .get(0)
+                .map(|ip_match| {
+                    let incorret_chars_on_bounds =
+                        [ip_match.start().wrapping_sub(1), ip_match.end()]
+                            .iter()
+                            .flat_map(|pos| log.chars().nth(*pos))
+                            .any(|c| c.is_alphanumeric() || c == '_' || c == '.');
+                    if incorret_chars_on_bounds {
+                        return ip_match.as_str().to_string();
+                    }
+
+                    IpAddr::from_str(ip_match.as_str())
+                        .map(|ip_addr| {
+                            // Probably good enough protection for this usecase
+                            let mut data_to_hash = vec![];
```

Comment:
nit: this probably doesn't have to be an allocation. We know that array of size 16+32 is big enough.